asmrun/roots.c

```diff
@@ -335,27 +335,89 @@ void caml_oldify_local_roots (void)
   if (caml_scan_roots_hook != NULL) (*caml_scan_roots_hook)(&caml_oldify_one);
 }
 
-/* Call [darken] on all roots */
+static mlsize_t incr_roots_i, incr_roots_j, incr_roots_sz, roots_count;
+static value *incr_roots_glob;
+uintnat caml_incremental_roots_count = 0;
+
+/* Call [caml_darken] on all roots, incrementally:
+   [caml_darken_all_roots_start] does the non-incremental part and
+   sets things up for [caml_darken_all_roots_slice].
+*/
+void caml_darken_all_roots_start (void)
+{
+  caml_do_roots (caml_darken, 0);
+  roots_count = 0;
+  for (incr_roots_i = 0; caml_globals[incr_roots_i] != 0; incr_roots_i++) {
+    for(incr_roots_glob = caml_globals[incr_roots_i]; *incr_roots_glob != 0;
+        incr_roots_glob++) {
+      incr_roots_sz = Wosize_val (*incr_roots_glob);
+      for (incr_roots_j = 0; incr_roots_j < incr_roots_sz; incr_roots_j++){
+        goto setup_done;
+      }
```

Comment:
The idea here is to have an exact copy of the loops found in `caml_do_roots`. Maybe I should add a comment to this effect.